File: Beecrowd/1195 Árvore Binária de Busca/main.py
class No:
    def __init__(self, conteudo):
        self.conteudo = conteudo
        self.esq = self.dir = None


class ArvoreBinaria:
    def __init__(self):
        self.raiz = None
        self.tam = 0

    def inserir(self, conteudo):
        for valor in conteudo:
            novo_conteudo = No(valor)
            self.raiz = self.inserir_aux(self.raiz, novo_conteudo)
            self.tam += 1

    def inserir_aux(self, atual, novo_no):
        if atual is None:
            return novo_no
        elif novo_no.conteudo < atual.conteudo:
            atual.esq = self.inserir_aux(atual.esq, novo_no)
        else:
            atual.dir = self.inserir_aux(atual.dir, novo_no)
        return atual

    def pre_ordem(self):
        global cont
        cont = 0
        self.pre_ordem_aux(self.raiz)

    def pre_ordem_aux(self, atual):
        global cont
        if atual is not None:
            if (cont < self.tam - 1):
                print(f"{atual.conteudo} ", end="")
            else:
                print(f"{atual.conteudo}", end="")
            cont += 1
            self.pre_ordem_aux(atual.esq)
            self.pre_ordem_aux(atual.dir)

    def em_ordem(self):
        global cont
        cont = 0
        self.em_ordem_aux(self.raiz)

    def em_ordem_aux(self, atual):
        global cont
        if atual is not None:
            self.em_ordem_aux(atual.esq)
            if (cont < self.tam - 1):
                print(f"{atual.conteudo} ", end="")
            else:
                print(f"{atual.conteudo}", end="")
            cont += 1
            self.em_ordem_aux(atual.dir)

    def pos_ordem(self):
        global cont
        cont = 0
        self.pos_ordem_aux(self.raiz)

    def pos_ordem_aux(self, atual):
        global cont
        if atual is not None:
            self.pos_ordem_aux(atual.esq)
            self.pos_ordem_aux(atual.dir)
            if (cont < self.tam - 1):
                print(f"{atual.conteudo} ", end="")
            else:
                print(f"{atual.conteudo}", end="")
            cont += 1


if __name__ == '__main__':
    c = int(input())

    for i in range(c):
        n = int(input())

        arv = ArvoreBinaria()
        valores = [int(_) for _ in input().split()[:n]]
        arv.inserir(valores)

        print(f"Case {i + 1}:")
        print("Pre.: ", end="")
        arv.pre_ordem()

        print("\nIn..: ", end="")
        arv.em_ordem()

        print("\nPost: ", end="")
        arv.pos_ordem()
        print("\n")

File: Beecrowd/1195 Árvore Binária de Busca/main.c
#include <stdio.h>
#include <stdlib.h>

typedef struct arv
{
    int valor;
    struct arv *esq;
    struct arv *dir;
} arv;

arv *cria();
int vazia(arv *a);
arv *insere(arv *a, int v);
void emOrdem(arv *a);
void preOrdem(arv *a);
void posOrdem(arv *a);
int contaElementos(arv *a);
void printArv(arv *arv, int i);

int cont, tam;

int main()
{
    arv *a;
    int c, n, valor;
    a = cria();

    scanf("%d", &c);

    for (int i = 0; i < c; i++)
    {
        scanf("%d", &n);

        for (int i = 0; i < n; i++)
        {
            scanf("%d", &valor);
            a = insere(a, valor);
        }

        printArv(a, i);

        a = (arv *)vazia(a);
    }

    return 0;
}

arv *cria()
{
    return NULL;
}

int vazia(arv *a)
{
    return (a == NULL);
}

arv *insere(arv *a, int v)
{
    if (a == NULL)
    {
        a = (arv *)malloc(sizeof(arv));
        a->valor = v;
        a->esq = a->dir = NULL;
    }
    else if (v < a->valor)
        a->esq = insere(a->esq, v);
    else
        a->dir = insere(a->dir, v);
    return a;
}

void emOrdem(arv *a)
{
    if (a != NULL)
    {
        emOrdem(a->esq);
        if (cont < tam - 1)
            printf("%d ", a->valor);
        else
            printf("%d", a->valor);
        cont++;
        emOrdem(a->dir);
    }
}

void preOrdem(arv *a)
{
    if (a != NULL)
    {
        if (cont < tam - 1)
            printf("%d ", a->valor);
        else
            printf("%d", a->valor);
        cont++;
        preOrdem(a->esq);
        preOrdem(a->dir);
    }
}

void posOrdem(arv *a)
{
    if (a != NULL)
    {
        posOrdem(a->esq);
        posOrdem(a->dir);
        if (cont < tam - 1)
            printf("%d ", a->valor);
        else
            printf("%d", a->valor);
        cont++;
    }
}

int contaElementos(arv *a)
{
    if (a == NULL)
        return 0;
    return 1 + contaElementos(a->esq) + contaElementos(a->dir);
}

void printArv(arv *arv, int i)
{
    tam = contaElementos(arv);
    cont = 0;
    printf("Case %d:\n", i + 1);
    printf("Pre.: ");
    preOrdem(arv);
    printf("\n");

    cont = 0;
    printf("In..: ");
    emOrdem(arv);
    printf("\n");

    cont = 0;
    printf("Post: ");
    posOrdem(arv);
    printf("\n\n");
}